File: algorithms/coding-interview-questions-blind75-lol/2/first_bad_version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        
        # hi = n
        # lo = 0
        # while hi > lo:
        #     mi = lo + (hi - lo) // 2
        #     if isBadVersion(mi):
        #         hi = mi
        #     else:
        #         lo = mi + 1
        # return hi

        b = n//2
        k = 0
        while b >= 1:
            while k+b <= n and not isBadVersion(b+k):
                k += b
            b //= 2
        
        return k+1

File: algorithms/coding-interview-questions-blind75-lol/2/first_bad_version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int hi = n;
        int lo = 1;

        int mi;
        while (hi > lo) {
            mi = lo + (hi - lo) / 2;
            if (isBadVersion(mi)) {
                hi = mi;
            } else {
                lo = mi + 1;
            }
        }
        return hi;

        // int k = 0;
        // for(int b = n/2; b >= 1; b /= 2) {
        //     while (k <= n-b && !isBadVersion(b+k)) k += b;
        // }
        // return k+1;
    }
};
